rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java

```diff
@@ -0,0 +1,220 @@
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * A circular buffer in redis. Head of the queue is the element at index 0 in redis.
+ */
+public class RedisCircularBuffer<ValueType> implements Queue<ValueType> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RedisCircularBuffer.class);
+
+    private final String key;
+    private final RDBI rdbi;
+    private final SerializationHelper<ValueType> serializationHelper;
+    private final int maxSize;
+
+    public RedisCircularBuffer(final RDBI rdbi,
+                               @Nonnull final String key,
+                               final int maxSize,
+                               SerializationHelper<ValueType> serializationHelper) {
+        this.rdbi = rdbi;
+        this.key = key;
+        this.maxSize = maxSize;
+        this.serializationHelper = serializationHelper;
+
+        Preconditions.checkNotNull(key, "A null value was supplied for 'key'.");
+    }
+
+    @Override
+    public int size() {
+        try (final Handle handle = rdbi.open()) {
+            final Long size = handle.jedis().llen(key);
+            if (size > Integer.MAX_VALUE) {
+                LOGGER.info("size of " + key + " exceeds integer max value. .size() just lied to you.");
+                return Integer.MAX_VALUE;
+            }
+            return size.intValue();
+        }
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        return containsAll(ImmutableList.of(o));
+    }
+
+    @Override
+    public Iterator<ValueType> iterator() {
+        throw new UnsupportedOperationException("iterator not supported by this circular buffer");
+    }
+
+    @Override
+    public Object[] toArray() {
+        throw new UnsupportedOperationException("toArray not supported by this circular buffer");
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a) {
+        throw new UnsupportedOperationException("toArray not supported by this circular buffer");
+    }
+
+    @Override
+    public boolean add(ValueType value) {
+        return addAll(ImmutableList.of(value));
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        throw new UnsupportedOperationException("removing a specific object is not supported by this circular buffer");
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> c) {
+        try (final Handle handle = rdbi.open()) {
+            final Pipeline pipeline = handle.jedis().pipelined();
+            Set<ValueType> currentSet = new HashSet<ValueType>();
+            final int size = handle.jedis().llen(key).intValue();
+            final Set<Response<String>> responses = Sets.newHashSetWithExpectedSize(size);
+
+            for (int i = 0; i < size; i++) {
+                responses.add(pipeline.lindex(key, i));
```

Comment:
same as peekAll, i think you should get all in one shot to redis, given how we plan to use this.

further, consider the fact that you could probably use peekAll as a replacement for most of what is currently in this method.